File: src/main/java/exnihilocreatio/json/CustomMeltableJson.java
package exnihilocreatio.json;

import com.google.gson.*;
import exnihilocreatio.registries.types.Meltable;
import exnihilocreatio.util.BlockInfo;

import java.lang.reflect.Type;

public class CustomMeltableJson implements JsonDeserializer<Meltable>, JsonSerializer<Meltable> {
    @Override
    public JsonElement serialize(Meltable src, Type typeOfSrc, JsonSerializationContext context) {
        JsonObject obj = new JsonObject();

        obj.addProperty("fluid", src.getFluid());
        obj.addProperty("amount", src.getAmount());

        if (!src.getTextureOverride().equals(BlockInfo.EMPTY)){
            obj.add("color", context.serialize(src.getTextureOverride(), BlockInfo.class));
        }

        return obj;
    }

    @Override
    public Meltable deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        JsonHelper helper = new JsonHelper(json);
        JsonObject obj = json.getAsJsonObject();

        String fluid = helper.getString("fluid");
        int amount = helper.getInteger("amount");

        BlockInfo textureOverride = BlockInfo.EMPTY;
        if (obj.has("textureOverride")) {
            textureOverride = context.deserialize(json.getAsJsonObject().get("textureOverride"), BlockInfo.class);
        }

        return new Meltable(fluid, amount, textureOverride);
    }
}

File: src/main/java/exnihilocreatio/json/CustomMeltableJson.kt
package exnihilocreatio.json

import com.google.gson.*
import exnihilocreatio.registries.types.Meltable
import exnihilocreatio.util.BlockInfo

import java.lang.reflect.Type

object CustomMeltableJson : JsonDeserializer<Meltable>, JsonSerializer<Meltable> {
    override fun serialize(src: Meltable, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        val obj = JsonObject()

        obj.addProperty("fluid", src.fluid)
        obj.addProperty("amount", src.amount)

        if (src.textureOverride != BlockInfo.EMPTY) {
            obj.add("color", context.serialize(src.textureOverride, BlockInfo::class.java))
        }

        return obj
    }

    @Throws(JsonParseException::class)
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): Meltable {
        val helper = JsonHelper(json)
        val obj = json.asJsonObject

        val fluid = helper.getString("fluid")
        val amount = helper.getInteger("amount")

        var textureOverride = BlockInfo.EMPTY
        if (obj.has("textureOverride")) {
            textureOverride = context.deserialize(json.asJsonObject.get("textureOverride"), BlockInfo::class.java)
        }

        return Meltable(fluid, amount, textureOverride)
    }
}
